File: cv code/tracking_v1.py
import numpy as np
import time

# --- 1. Kalman Filter Implementation ---
# This class remains IDENTICAL as its purpose is purely state estimation.
class KalmanFilter:
    """
    A 2D Kalman Filter for tracking an object's position and velocity.
    State vector: [x, y, vx, vy] (position and velocity in x and y directions)
    Measurement vector: [x, y] (observed position from YOLO)
    """
    def __init__(self, dt, process_noise_cov, measurement_noise_cov, initial_state, initial_covariance):
        self.dt = dt
        self.A = np.array([
            [1, 0, dt, 0],
            [0, 1, 0, dt],
            [0, 0, 1, 0],
            [0, 0, 0, 1]
        ])
        self.H = np.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0]
        ])
        self.Q = np.eye(4) * process_noise_cov
        self.R = np.eye(2) * measurement_noise_cov
        self.x_hat = initial_state
        self.P = initial_covariance

    def predict(self):
        self.x_hat = np.dot(self.A, self.x_hat)
        self.P = np.dot(np.dot(self.A, self.P), self.A.T) + self.Q

    def update(self, measurement):
        y = measurement - np.dot(self.H, self.x_hat)
        S = np.dot(np.dot(self.H, self.P), self.H.T) + self.R
        K = np.dot(np.dot(self.P, self.H.T), np.linalg.inv(S))
        self.x_hat = self.x_hat + np.dot(K, y)
        self.P = np.dot((np.eye(self.A.shape[0]) - np.dot(K, self.H)), self.P)

    def get_estimated_position(self):
        return self.x_hat[0], self.x_hat[1]

# --- 2. PID Controller Implementation ---
# This class also remains IDENTICAL. The interpretation of its output changes.
class PIDController:
    """
    A generic PID controller for generating control signals.
    """
    def __init__(self, kp, ki, kd, output_min, output_max, integral_max=None):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.output_min = output_min
        self.output_max = output_max
        self.integral_max = integral_max

        self._previous_error = 0.0
        self._integral = 0.0
        self._last_time = None

    def calculate(self, setpoint, process_variable):
        current_time = time.time()
        if self._last_time is None:
            self._last_time = current_time
            return 0.0

        dt = current_time - self._last_time
        if dt == 0:
            return self.output_min if self._previous_error < 0 else self.output_max

        error = setpoint - process_variable

        p_term = self.kp * error

        self._integral += error * dt
        if self.integral_max is not None:
            self._integral = max(min(self._integral, self.integral_max), -self.integral_max)
        i_term = self.ki * self._integral

        d_term = self.kd * ((error - self._previous_error) / dt)

        output = p_term + i_term + d_term
        output = max(self.output_min, min(output, self.output_max))

        self._previous_error = error
        self._last_time = current_time

        return output

    def reset(self):
        self._previous_error = 0.0
        self._integral = 0.0
        self._last_time = None

# --- 3. ESP32 Servo Control (Conceptual MicroPython) ---
# This section remains conceptually the same but explicitly targets a gimbal.

# IMPORTANT: This part requires MicroPython firmware on your ESP32
# and the 'machine' module. It cannot be run directly as standard Python.


from machine import Pin, PWM
import time

class ESP32GimbalControl:
    def __init__(self, pitch_pin_num, yaw_pin_num, freq=50):
        self.freq = freq

        self.pitch_pin = Pin(pitch_pin_num)
        self.pitch_pwm = PWM(self.pitch_pin, freq=self.freq)

        self.yaw_pin = Pin(yaw_pin_num)
        self.yaw_pwm = PWM(self.yaw_pin, freq=self.freq)

        # Duty cycle values for 0.5ms (0 deg) to 2.5ms (180 deg) pulse for 50Hz PWM
        self.min_duty = self.ms_to_duty_cycle(0.5)
        self.max_duty = self.ms_to_duty_cycle(2.5)
        self.range_duty = self.max_duty - self.min_duty

    def ms_to_duty_cycle(self, ms_pulse):
        return int((ms_pulse / (1000 / self.freq)) * 1023) # ESP32 duty cycle resolution

    def set_servo_angle(self, servo_pwm, angle_degrees):
        angle_degrees = max(0, min(angle_degrees, 180)) # Clamp to 0-180 for standard servos
        duty = self.min_duty + (angle_degrees / 180.0) * self.range_duty
        servo_pwm.duty(int(duty))

    def move_gimbal(self, pitch_angle, yaw_angle):
        # Set pitch and yaw servo angles
        # Note: You might need to adjust which servo receives which angle
        # and the direction (e.g., pitch_angle vs 180-pitch_angle) based on your gimbal.
        self.set_servo_angle(self.pitch_pwm, pitch_angle)
        self.set_servo_angle(self.yaw_pwm, yaw_angle)

    def deinit(self):
        self.pitch_pwm.deinit()
        self.yaw_pwm.deinit()


# --- 4. Main Tracking Loop Simulation (for demonstration) ---
# The logic here is also largely the same, but the 'target' is now the gimbal.

if __name__ == "__main__":
    # --- Configuration Parameters ---
    DT = 0.1 # seconds (e.g., 10 frames per second)

    # Kalman Filter Parameters
    PROCESS_NOISE_COV = 0.1
    MEASUREMENT_NOISE_COV = 10.0
    INITIAL_STATE = np.array([0.0, 0.0, 0.0, 0.0])
    INITIAL_COVARIANCE = np.eye(4) * 100.0

    # PID Controller Parameters
    # These gains need to be tuned for your specific GIMBAL and its servos.
    # The output range now represents degrees of *correction* for the gimbal.
    # For instance, an output of 10 means rotate the gimbal 10 degrees.
    # Assuming your gimbal can move +/- 30 degrees from its center.
    PID_OUTPUT_MAX_DEGREE = 30 # Max degrees the PID asks the gimbal to move from its current position
    PID_OUTPUT_MIN_DEGREE = -30

    KP_PITCH = 0.1 # Tune for camera's vertical movement
    KI_PITCH = 0.01
    KD_PITCH = 0.005
    INTEGRAL_MAX_PITCH = 10

    KP_YAW = 0.1 # Tune for camera's horizontal movement
    KI_YAW = 0.01
    KD_YAW = 0.005
    INTEGRAL_MAX_YAW = 10

    # Camera Frame Center (Target for PID)
    CAMERA_CENTER_X = 320 # Example: For a 640x480 resolution
    CAMERA_CENTER_Y = 240 # Example: For a 640x480 resolution

    # Servo Angle Mapping for Gimbal
    # This assumes the gimbal servos are commanded in absolute angles (0-180 degrees).
    # You'll need to know the *current* angle of your gimbal.
    # For a simple control, you might adjust the current angle by the PID output.
    # Let's assume an initial gimbal position and current angles
    # In a real system, you might read these from encoders or assume a home position.
    current_gimbal_pitch_angle = 90.0 # Start at 90 degrees (center)
    current_gimbal_yaw_angle = 90.0   # Start at 90 degrees (center)

    def calculate_new_gimbal_angle(current_angle, pid_output, servo_min_limit=0, servo_max_limit=180):
        """
        Calculates the new target angle for a gimbal servo based on PID output.
        PID output is a *correction* in degrees.
        Positive PID output on YAW means move camera to the right (increase yaw angle).
        Positive PID output on PITCH means move camera up (decrease pitch angle, assuming image Y increases downwards).
        You'll need to confirm these directions with your physical setup.
        """
        # For X-axis (Yaw), if PID output is positive, we want to move the camera right to center the object.
        # This typically means increasing the yaw servo angle.
        # For Y-axis (Pitch), if PID output is positive, we want to move the camera up to center the object.
        # If your image Y increases downwards, moving 'up' means decreasing the pitch servo angle.
        # THIS IS A CRITICAL MAPPING that depends on your gimbal's orientation!
        # Example mapping (adjust signs based on your gimbal):
        # new_angle = current_angle + pid_output # For yaw where increasing angle moves right
        # new_angle = current_angle - pid_output # For pitch where decreasing angle moves up

        # Let's assume:
        # Yaw PID output: positive means rotate camera to the right (increase servo angle)
        # Pitch PID output: positive means rotate camera down (increase servo angle, if Y increases down)
        # IF YOU WANT TO MOVE CAMERA UP FOR POSITIVE PID, you need to subtract.
        # This example assumes positive PID output increases servo angle.
        # Adjust for YOUR gimbal's actual behavior:
        
        # Example: pid_output > 0 means object is right/down of center, need to move gimbal right/down
        # So, increase yaw angle to look right, increase pitch angle to look down.
        new_angle = current_angle + pid_output
        return max(servo_min_limit, min(servo_max_limit, new_angle))


    # --- Initialize Components ---
    kalman_filter = KalmanFilter(DT, PROCESS_NOISE_COV, MEASUREMENT_NOISE_COV,
                                 INITIAL_STATE, INITIAL_COVARIANCE)

    # PID for Yaw control (horizontal error, X-axis -> controls gimbal yaw)
    pid_yaw = PIDController(KP_YAW, KI_YAW, KD_YAW, PID_OUTPUT_MIN_DEGREE, PID_OUTPUT_MAX_DEGREE, INTEGRAL_MAX_YAW)
    # PID for Pitch control (vertical error, Y-axis -> controls gimbal pitch)
    pid_pitch = PIDController(KP_PITCH, KI_PITCH, KD_PITCH, PID_OUTPUT_MIN_DEGREE, PID_OUTPUT_MAX_DEGREE, INTEGRAL_MAX_PITCH)

    # ESP32 Gimbal Control (Conceptual - would run on ESP32)
    # gimbal_controller = ESP32GimbalControl(pitch_pin_num=2, yaw_pin_num=4) # Example pins

    print("Starting camera gimbal tracking simulation...")
    print("Press Ctrl+C to stop.")

    # --- Simulation Loop ---
    simulated_true_x = 0
    simulated_true_y = 0
    simulated_vx = 5 # pixels/sec
    simulated_vy = 3 # pixels/sec
    frame_count = 0

    try:
        while True:
            start_time = time.time()

            # --- 1. Simulate YOLO Measurement (replace with actual YOLO output) ---
            simulated_true_x += simulated_vx * DT
            simulated_true_y += simulated_vy * DT

            yolo_measured_x = simulated_true_x + np.random.normal(0, np.sqrt(MEASUREMENT_NOISE_COV))
            yolo_measured_y = simulated_true_y + np.random.normal(0, np.sqrt(MEASUREMENT_NOISE_COV))
            current_measurement = np.array([yolo_measured_x, yolo_measured_y])

            # --- 2. Kalman Filter Prediction and Update ---
            kalman_filter.predict()
            kalman_filter.update(current_measurement)
            estimated_x, estimated_y = kalman_filter.get_estimated_position()

            print(f"\n--- Frame {frame_count} ---")
            print(f"Simulated True Pos: ({simulated_true_x:.2f}, {simulated_true_y:.2f})")
            print(f"YOLO Measured Pos:  ({yolo_measured_x:.2f}, {yolo_measured_y:.2f})")
            print(f"Kalman Estimated Pos: ({estimated_x:.2f}, {estimated_y:.2f})")

            # --- 3. PID Control Calculation ---
            # PID calculates the *correction* needed in degrees for the gimbal.
            # Error: Camera_Center - Estimated_Object_Position
            yaw_pid_output = pid_yaw.calculate(CAMERA_CENTER_X, estimated_x)
            pitch_pid_output = pid_pitch.calculate(CAMERA_CENTER_Y, estimated_y)

            print(f"Yaw PID Correction: {yaw_pid_output:.2f} degrees (Target X: {CAMERA_CENTER_X})")
            print(f"Pitch PID Correction: {pitch_pid_output:.2f} degrees (Target Y: {CAMERA_CENTER_Y})")

            # --- 4. Gimbal Servo Control ---
            # Update current gimbal angles based on PID output
            # This is where the core change happens: PID output directly influences the gimbal's current angle.
            target_gimbal_yaw_angle = calculate_new_gimbal_angle(current_gimbal_yaw_angle, yaw_pid_output)
            target_gimbal_pitch_angle = calculate_new_gimbal_angle(current_gimbal_pitch_angle, pitch_pid_output)

            # For simulation, update the 'current' angle for the next iteration
            current_gimbal_yaw_angle = target_gimbal_yaw_angle
            current_gimbal_pitch_angle = target_gimbal_pitch_angle

            print(f"Target Gimbal Yaw Angle: {target_gimbal_yaw_angle:.2f} degrees")
            print(f"Target Gimbal Pitch Angle: {target_gimbal_pitch_angle:.2f} degrees")

            # In a real ESP32 setup, you would call:
            # gimbal_controller.move_gimbal(target_gimbal_pitch_angle, target_gimbal_yaw_angle)

            # --- Maintain Loop Timing ---
            end_time = time.time()
            elapsed_time = end_time - start_time
            sleep_time = DT - elapsed_time
            if sleep_time > 0:
                time.sleep(sleep_time)
            else:
                print(f"Warning: Loop took longer than DT ({elapsed_time:.4f}s vs {DT}s)")

            frame_count += 1

    except KeyboardInterrupt:
        print("\nSimulation stopped by user.")
        # In a real ESP32 setup, you would de-initialize servos:
        # gimbal_controller.deinit()

File: cv code/tracking_v1.c
#include <stdio.h>
#include <math.h>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "esp_log.h"
#include "driver/ledc.h" // For PWM control of servos
#include "esp_timer.h"   // For high-resolution timing

// --- Configuration Parameters ---
#define DT 0.1f // seconds (e.g., 10 frames per second)

// Kalman Filter Parameters
#define PROCESS_NOISE_COV 0.1f
#define MEASUREMENT_NOISE_COV 10.0f

// PID Controller Parameters
#define PID_OUTPUT_MAX_DEGREE 30.0f
#define PID_OUTPUT_MIN_DEGREE -30.0f

#define KP_PITCH 0.1f
#define KI_PITCH 0.01f
#define KD_PITCH 0.005f
#define INTEGRAL_MAX_PITCH 10.0f

#define KP_YAW 0.1f
#define KI_YAW 0.01f
#define KD_YAW 0.005f
#define INTEGRAL_MAX_YAW 10.0f

// Camera Frame Center (Target for PID)
#define CAMERA_CENTER_X 320.0f // Example: For a 640x480 resolution
#define CAMERA_CENTER_Y 240.0f // Example: For a 640x480 resolution

// Servo Pin Definitions (Adjust these to your ESP32 GPIOs)
// Use GPIOs that support LEDC (PWM). Common choices: 0, 2, 4, 5, 12-19, 21-23, 25-27.
#define PITCH_SERVO_GPIO 13
#define YAW_SERVO_GPIO 12

// Servo PWM Configuration
#define SERVO_PWM_FREQ 50      // 50 Hz for standard servos
#define LEDC_TIMER LEDC_TIMER_0
#define LEDC_MODE LEDC_LOW_SPEED_MODE
#define LEDC_DUTY_RES LEDC_TIMER_10_BIT // 10-bit resolution (0-1023)

// Servo Pulse Widths in microseconds (uS)
// These values need to be calibrated for your specific servos.
// Standard servos: 0 deg ~500uS, 90 deg ~1500uS, 180 deg ~2500uS
#define SERVO_MIN_PULSE_US 500
#define SERVO_MAX_PULSE_US 2500

static const char *TAG = "GIMBAL_TRACKING";

// --- 1. Kalman Filter Implementation ---

// Matrix multiplication helper (C = A * B)
void mat_mul(float *A, float *B, float *C, int r1, int c1, int r2, int c2) {
    for (int i = 0; i < r1; i++) {
        for (int j = 0; j < c2; j++) {
            C[i * c2 + j] = 0;
            for (int k = 0; k < c1; k++) {
                C[i * c2 + j] += A[i * c1 + k] * B[k * c2 + j];
            }
        }
    }
}

// Matrix addition helper (C = A + B)
void mat_add(float *A, float *B, float *C, int r, int c) {
    for (int i = 0; i < r * c; i++) {
        C[i] = A[i] + B[i];
    }
}

// Matrix subtraction helper (C = A - B)
void mat_sub(float *A, float *B, float *C, int r, int c) {
    for (int i = 0; i < r * c; i++) {
        C[i] = A[i] - B[i];
    }
}

// Matrix transpose helper (B = A_T)
void mat_transpose(float *A, float *B, int r, int c) {
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            B[j * r + i] = A[i * c + j];
        }
    }
}

// 2x2 Matrix Inverse (only for 2x2, as needed for S in Kalman)
int mat_inv_2x2(float *A, float *B) {
    float det = A[0] * A[3] - A[1] * A[2];
    if (fabs(det) < 1e-6) { // Check for singularity
        ESP_LOGE(TAG, "Matrix is singular, cannot invert.");
        return -1;
    }
    float inv_det = 1.0f / det;
    B[0] = A[3] * inv_det;
    B[1] = -A[1] * inv_det;
    B[2] = -A[2] * inv_det;
    B[3] = A[0] * inv_det;
    return 0;
}

typedef struct {
    float dt;
    float A[4*4]; // State transition matrix
    float H[2*4]; // Measurement matrix
    float Q[4*4]; // Process noise covariance
    float R[2*2]; // Measurement noise covariance
    float x_hat[4]; // State estimate [x, y, vx, vy]
    float P[4*4]; // Estimate covariance
} KalmanFilter;

void kalman_filter_init(KalmanFilter *kf, float dt, float process_noise_cov, float measurement_noise_cov,
                        float initial_x, float initial_y, float initial_vx, float initial_vy) {
    kf->dt = dt;

    // Initialize A (State transition matrix)
    float A_init[4*4] = {
        1, 0, dt, 0,
        0, 1, 0, dt,
        0, 0, 1, 0,
        0, 0, 0, 1
    };
    memcpy(kf->A, A_init, sizeof(A_init));

    // Initialize H (Measurement matrix)
    float H_init[2*4] = {
        1, 0, 0, 0,
        0, 1, 0, 0
    };
    memcpy(kf->H, H_init, sizeof(H_init));

    // Initialize Q (Process noise covariance)
    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            kf->Q[i*4 + j] = (i == j) ? process_noise_cov : 0;
        }
    }

    // Initialize R (Measurement noise covariance)
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            kf->R[i*2 + j] = (i == j) ? measurement_noise_cov : 0;
        }
    }

    // Initial state estimate
    kf->x_hat[0] = initial_x;
    kf->x_hat[1] = initial_y;
    kf->x_hat[2] = initial_vx;
    kf->x_hat[3] = initial_vy;

    // Initial estimate covariance (P)
    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            kf->P[i*4 + j] = (i == j) ? 100.0f : 0; // High initial uncertainty
        }
    }
}

void kalman_filter_predict(KalmanFilter *kf) {
    float temp_x_hat[4];
    mat_mul(kf->A, kf->x_hat, temp_x_hat, 4, 4, 4, 1); // x_hat = A * x_hat
    memcpy(kf->x_hat, temp_x_hat, sizeof(temp_x_hat));

    float A_T[4*4];
    mat_transpose(kf->A, A_T, 4, 4);
    float temp_P[4*4];
    mat_mul(kf->A, kf->P, temp_P, 4, 4, 4, 4); // temp_P = A * P
    float new_P[4*4];
    mat_mul(temp_P, A_T, new_P, 4, 4, 4, 4); // new_P = temp_P * A_T
    mat_add(new_P, kf->Q, kf->P, 4, 4); // P = new_P + Q
}

void kalman_filter_update(KalmanFilter *kf, float measurement_x, float measurement_y) {
    float measurement[2] = {measurement_x, measurement_y};

    float H_x_hat[2];
    mat_mul(kf->H, kf->x_hat, H_x_hat, 2, 4, 4, 1); // H * x_hat

    float y[2];
    mat_sub(measurement, H_x_hat, y, 2, 1); // y = z - H * x_hat

    float H_T[4*2];
    mat_transpose(kf->H, H_T, 2, 4);

    float P_H_T[4*2];
    mat_mul(kf->P, H_T, P_H_T, 4, 4, 4, 2); // P * H_T

    float H_P_H_T[2*2];
    mat_mul(kf->H, P_H_T, H_P_H_T, 2, 4, 4, 2); // H * P * H_T

    float S[2*2];
    mat_add(H_P_H_T, kf->R, S, 2, 2); // S = H * P * H_T + R

    float S_inv[2*2];
    if (mat_inv_2x2(S, S_inv) != 0) {
        // Handle error, e.g., return or log
        return;
    }

    float K[4*2];
    mat_mul(P_H_T, S_inv, K, 4, 2, 2, 2); // K = P * H_T * S_inv

    float K_y[4];
    mat_mul(K, y, K_y, 4, 2, 2, 1); // K * y
    mat_add(kf->x_hat, K_y, kf->x_hat, 4, 1); // x_hat = x_hat + K * y

    float I_minus_KH[4*4];
    float K_H[4*4];
    mat_mul(K, kf->H, K_H, 4, 2, 2, 4); // K * H
    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            I_minus_KH[i*4 + j] = (i == j) ? (1.0f - K_H[i*4 + j]) : (-K_H[i*4 + j]);
        }
    }
    float new_P_updated[4*4];
    mat_mul(I_minus_KH, kf->P, new_P_updated, 4, 4, 4, 4); // P = (I - K * H) * P
    memcpy(kf->P, new_P_updated, sizeof(new_P_updated));
}

// --- 2. PID Controller Implementation ---

typedef struct {
    float kp;
    float ki;
    float kd;
    float output_min;
    float output_max;
    float integral_max;

    float previous_error;
    float integral;
    int64_t last_time_us; // Use microseconds for precision
} PIDController;

void pid_controller_init(PIDController *pid, float kp, float ki, float kd,
                         float output_min, float output_max, float integral_max) {
    pid->kp = kp;
    pid->ki = ki;
    pid->kd = kd;
    pid->output_min = output_min;
    pid->output_max = output_max;
    pid->integral_max = integral_max;

    pid->previous_error = 0.0f;
    pid->integral = 0.0f;
    pid->last_time_us = 0; // Will be set on first calculate call
}

float pid_controller_calculate(PIDController *pid, float setpoint, float process_variable) {
    int64_t current_time_us = esp_timer_get_time();
    if (pid->last_time_us == 0) {
        pid->last_time_us = current_time_us;
        return 0.0f; // Return 0 on first call as dt is not defined
    }

    float dt = (float)(current_time_us - pid->last_time_us) / 1000000.0f; // Convert uS to seconds
    pid->last_time_us = current_time_us;

    if (dt == 0) {
        return (pid->previous_error < 0) ? pid->output_min : pid->output_max;
    }

    float error = setpoint - process_variable;

    // Proportional term
    float p_term = pid->kp * error;

    // Integral term
    pid->integral += error * dt;
    if (pid->integral_max > 0) { // Only clamp if integral_max is set
        pid->integral = fmaxf(fminf(pid->integral, pid->integral_max), -pid->integral_max);
    }
    float i_term = pid->ki * pid->integral;

    // Derivative term
    float d_term = pid->kd * ((error - pid->previous_error) / dt);

    // Total output
    float output = p_term + i_term + d_term;

    // Clamp output to min/max values
    output = fmaxf(pid->output_min, fminf(output, pid->output_max));

    // Store current error for the next iteration
    pid->previous_error = error;

    return output;
}

void pid_controller_reset(PIDController *pid) {
    pid->previous_error = 0.0f;
    pid->integral = 0.0f;
    pid->last_time_us = 0;
}

// --- 3. ESP32 Gimbal Servo Control ---

// Function to convert angle (0-180) to PWM duty cycle
// Duty cycle is calculated based on pulse width (uS) for a given frequency and resolution.
uint32_t angle_to_duty(float angle_degrees) {
    // Clamp angle to valid range
    angle_degrees = fmaxf(0.0f, fminf(180.0f, angle_degrees));

    // Calculate pulse width in microseconds
    // Map 0-180 degrees to SERVO_MIN_PULSE_US to SERVO_MAX_PULSE_US
    float pulse_us = SERVO_MIN_PULSE_US + (angle_degrees / 180.0f) * (SERVO_MAX_PULSE_US - SERVO_MIN_PULSE_US);

    // Calculate duty cycle from pulse width
    // Duty cycle = (pulse_us / (1,000,000 uS / SERVO_PWM_FREQ)) * (2^LEDC_DUTY_RES - 1)
    // For 50Hz, period = 20,000 uS
    uint32_t duty = (uint32_t)((pulse_us / (1000000.0f / SERVO_PWM_FREQ)) * (1 << LEDC_DUTY_RES));
    return duty;
}

// Initialize LEDC peripheral for a specific channel
void servo_init(ledc_channel_t channel, int gpio_num) {
    ledc_timer_config_t ledc_timer = {
        .speed_mode = LEDC_MODE,
        .timer_num = LEDC_TIMER,
        .duty_resolution = LEDC_DUTY_RES,
        .freq_hz = SERVO_PWM_FREQ,
        .clk_cfg = LEDC_AUTO_CLK,
    };
    ESP_ERROR_CHECK(ledc_timer_config(&ledc_timer));

    ledc_channel_config_t ledc_channel = {
        .speed_mode = LEDC_MODE,
        .channel = channel,
        .timer_sel = LEDC_TIMER,
        .intr_type = LEDC_INTR_DISABLE,
        .gpio_num = gpio_num,
        .duty = 0, // Set duty to 0 initially
        .hpoint = 0,
    };
    ESP_ERROR_CHECK(ledc_channel_config(&ledc_channel));

    ESP_LOGI(TAG, "Servo on GPIO %d initialized for LEDC channel %d", gpio_num, channel);
}

// Set servo angle for a specific channel
void servo_set_angle(ledc_channel_t channel, float angle_degrees) {
    uint32_t duty = angle_to_duty(angle_degrees);
    ESP_ERROR_CHECK(ledc_set_duty(LEDC_MODE, channel, duty));
    ESP_ERROR_CHECK(ledc_update_duty(LEDC_MODE, channel));
}

// --- Main Application ---

// Global variables for current gimbal angles (can be updated by PID)
volatile float current_gimbal_pitch_angle = 90.0f; // Start at 90 degrees (center)
volatile float current_gimbal_yaw_angle = 90.0f;   // Start at 90 degrees (center)

// Function to calculate new gimbal angle based on PID output
float calculate_new_gimbal_angle(float current_angle, float pid_output, float servo_min_limit, float servo_max_limit) {
    // PID output is a *correction* in degrees.
    // Positive PID output on YAW means object is right of center, need to move gimbal right (increase yaw angle).
    // Positive PID output on PITCH means object is below center, need to move gimbal down (increase pitch angle).
    // Adjust signs based on your physical gimbal mounting and servo direction.
    float new_angle = current_angle + pid_output;
    return fmaxf(servo_min_limit, fminf(servo_max_limit, new_angle));
}

void app_main(void) {
    ESP_LOGI(TAG, "Starting ESP32 Camera Gimbal Tracking System...");

    // Initialize Kalman Filter
    KalmanFilter kf;
    kalman_filter_init(&kf, DT, PROCESS_NOISE_COV, MEASUREMENT_NOISE_COV,
                       0.0f, 0.0f, 0.0f, 0.0f); // Initial state [x,y,vx,vy]

    // Initialize PID Controllers
    PIDController pid_yaw, pid_pitch;
    pid_controller_init(&pid_yaw, KP_YAW, KI_YAW, KD_YAW,
                        PID_OUTPUT_MIN_DEGREE, PID_OUTPUT_MAX_DEGREE, INTEGRAL_MAX_YAW);
    pid_controller_init(&pid_pitch, KP_PITCH, KI_PITCH, KD_PITCH,
                         PID_OUTPUT_MIN_DEGREE, PID_OUTPUT_MAX_DEGREE, INTEGRAL_MAX_PITCH);

    // Initialize Servos
    servo_init(LEDC_CHANNEL_0, PITCH_SERVO_GPIO); // Pitch servo on LEDC Channel 0
    servo_init(LEDC_CHANNEL_1, YAW_SERVO_GPIO);   // Yaw servo on LEDC Channel 1

    // Set initial servo positions
    servo_set_angle(LEDC_CHANNEL_0, current_gimbal_pitch_angle);
    servo_set_angle(LEDC_CHANNEL_1, current_gimbal_yaw_angle);
    vTaskDelay(pdMS_TO_TICKS(1000)); // Give servos time to move to initial position

    // --- Simulation Variables (Replace with actual YOLO input) ---
    float simulated_true_x = 0.0f;
    float simulated_true_y = 0.0f;
    float simulated_vx = 5.0f; // pixels/sec
    float simulated_vy = 3.0f; // pixels/sec
    uint32_t frame_count = 0;
    int64_t loop_start_time_us;
    int64_t loop_end_time_us;
    int64_t elapsed_time_us;
    int64_t sleep_time_us;

    // Main Tracking Loop
    while (1) {
        loop_start_time_us = esp_timer_get_time();

        // --- REPLACE THIS SECTION WITH YOUR ACTUAL YOLO DATA INPUT ---
        // This is where you would receive (x, y) coordinates from your YOLO system.
        // For example, reading from UART, Wi-Fi, or a camera module.
        // For now, we simulate a noisy measurement of a moving object.
        simulated_true_x += simulated_vx * DT;
        simulated_true_y += simulated_vy * DT;

        float yolo_measured_x = simulated_true_x + ((float)rand() / RAND_MAX - 0.5f) * 2.0f * sqrtf(MEASUREMENT_NOISE_COV);
        float yolo_measured_y = simulated_true_y + ((float)rand() / RAND_MAX - 0.5f) * 2.0f * sqrtf(MEASUREMENT_NOISE_COV);
        // -----------------------------------------------------------

        // Kalman Filter Prediction and Update
        kalman_filter_predict(&kf);
        kalman_filter_update(&kf, yolo_measured_x, yolo_measured_y);
        float estimated_x = kf.x_hat[0];
        float estimated_y = kf.x_hat[1];

        // PID Control Calculation
        float yaw_pid_output = pid_yaw.calculate(&pid_yaw, CAMERA_CENTER_X, estimated_x);
        float pitch_pid_output = pid_pitch.calculate(&pid_pitch, CAMERA_CENTER_Y, estimated_y);

        // Calculate new gimbal angles
        current_gimbal_yaw_angle = calculate_new_gimbal_angle(current_gimbal_yaw_angle, yaw_pid_output, 0.0f, 180.0f);
        current_gimbal_pitch_angle = calculate_new_gimbal_angle(current_gimbal_pitch_angle, pitch_pid_output, 0.0f, 180.0f);

        // Move Gimbal Servos
        servo_set_angle(LEDC_CHANNEL_0, current_gimbal_pitch_angle);
        servo_set_angle(LEDC_CHANNEL_1, current_gimbal_yaw_angle);

        // --- Debug Logging (can be disabled for performance) ---
        ESP_LOGI(TAG, "Frame %lu: YOLO(%.2f,%.2f) Est(%.2f,%.2f) PID Y/P(%.2f,%.2f) Gimbal Y/P(%.2f,%.2f)",
                 frame_count, yolo_measured_x, yolo_measured_y, estimated_x, estimated_y,
                 yaw_pid_output, pitch_pid_output, current_gimbal_yaw_angle, current_gimbal_pitch_angle);
        // -------------------------------------------------------

        // Maintain Loop Timing
        loop_end_time_us = esp_timer_get_time();
        elapsed_time_us = loop_end_time_us - loop_start_time_us;
        sleep_time_us = (int64_t)(DT * 1000000) - elapsed_time_us;

        if (sleep_time_us > 0) {
            vTaskDelay(pdUS_TO_TICKS(sleep_time_us)); // Delay in microseconds
        } else {
            ESP_LOGW(TAG, "Loop took longer than DT (%.2fms vs %.2fms)",
                     (float)elapsed_time_us / 1000.0f, DT * 1000.0f);
        }

        frame_count++;
    }
}
